master/internal/core_trial.go

```diff
@@ -1,14 +1,57 @@
 package internal
 
 import (
+	"net/http"
+	"strconv"
+
 	"github.com/labstack/echo/v4"
+	"github.com/pkg/errors"
+
+	"github.com/determined-ai/determined/master/internal/context"
+	"github.com/determined-ai/determined/master/internal/db"
+	expauth "github.com/determined-ai/determined/master/internal/experiment"
 )
 
+func echoCanGetTrial(c echo.Context, m *Master, trialID string) error {
+	id, err := strconv.Atoi(trialID)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, "trial ID must be numeric got %s", trialID)
+	}
+
+	curUser := c.(*context.DetContext).MustGetUser()
+	trialNotFound := echo.NewHTTPError(http.StatusNotFound, "trial not found: %d", id)
+	exp, err := m.db.ExperimentWithoutConfigByTrialID(id)
+	if errors.Is(err, db.ErrNotFound) {
+		return trialNotFound
+	} else if err != nil {
+		return err
+	}
+	var ok bool
+	if ok, err = expauth.AuthZProvider.Get().CanGetExperiment(curUser, exp); err != nil {
+		return err
+	} else if !ok {
+		return trialNotFound
+	}
+
+	if err = expauth.AuthZProvider.Get().CanGetExperimentArtifacts(curUser, exp); err != nil {
+		return echo.NewHTTPError(http.StatusForbidden, err.Error())
+	}
+	return nil
+}
+
 // TODO(ilia): These APIs are deprecated and will be removed in a future release.
 func (m *Master) getTrial(c echo.Context) (interface{}, error) {
+	if err := echoCanGetTrial(c, m, c.Param("trial_id")); err != nil {
```

Comment:
reminder to myself: we should remove both of these in the next major version: 0.20 or 0.21 at the latest.